uast/uast-kotlin-base/src/org/jetbrains/uast/kotlin/psi/UastKotlinPsiParameter.kt

```diff
@@ -29,17 +27,23 @@ class UastKotlinPsiParameter(
     ktParameter: KtParameter
 ) : UastKotlinPsiParameterBase<KtParameter>(name, type, parent, ktParameter, language, isVarArgs, ktDefaultValue) {
     companion object {
-        fun create(parameter: KtParameter, parent: PsiElement, containingElement: UElement, index: Int): PsiParameter {
+        fun create(
+            baseKotlinUastResolveProviderService: BaseKotlinUastResolveProviderService,
```

Comment:
I don't get why we pass `BaseKotlinUastResolveProviderService` everyhere instead of getting it as a Service from project of PsiElement

Reply:
As described in this commit message: https://github.com/JetBrains/intellij-kotlin/commit/a4a79a9b20e84016a26af7ca8d292295eac903d9, `project` of `PsiElement` may not be available. I remember seeing that occasionally while working in `kotlin` repo, but apparently much more frequent in `intellij-kotlin` repo, maybe src/test differences I didn't observe previously.

Since we enforce that base service to be present, for such case where `project` of `PsiElement` is absent, we can still provide the service instance where conversion is happening. That's my understanding, perhaps @darthorimar can add more.

Reply:
But in the current case it is avaliable, you have non-null `PsiElement` and `PsiElement.getProject` is also strcitcly not-null. On other hand we could consider making `BaseKotlinUastResolveProviderService` an application service and not depend on project at all. I belive there is no case when we could want to configure different `BaseKotlinUastResolveProviderService` for different projects

Reply:
I see. Yeah, for `UastKotlinPsi*` cases, `PsiElement` exists, and it seems to work if I retrieve the service from the `project` of that `PsiElement` (if that service is needed for some other purposes).

However, being an application service sounds more convincing, so I tried that, but got stuck (thanks to this [place](https://github.com/JetBrains/intellij-kotlin/pull/82#discussion_r667297137)). Then, realized that it will make several conflicts with piles of commits/PRs I already have: #79 and #80. So I made a checkpoint (#81), and attempted it on top of that again. If you don't mind, we can continue this discussion on #82, which indeed simplifies all the place of passing this base service.

Reply:
I don't mind, we can solve it in any other PR. Sorry I didn't meant to block this PR from merging.

Reply:
No problem! Thank you for your understanding. As mentioned, we can resolve this issue in that PR.